_posts/2018-12-06.Rust-1.31.md

```diff
@@ -0,0 +1,338 @@
+first iteration of "Rust 2018," but there's more than just that! This is going to be a long
+post, so here's a table of contents:
+
+* [Rust 2018](#rust-2018)
+  * [Non-lexical lifetimes](#non-lexical-lifetimes)
+  * [Module system changes](#module-system-changes)
+* [More lifetime elision rules](#more-lifetime-elision-rules)
+* [`const fn`](#const-fn)
+* [Tool Lints](#tool-lints)
+* [Library stabilizations](#library-stabilizations)
+* [Cargo features](#cargo-features)
+* [Contributors](#contributors-to-1310)
+
+### Rust 2018
+
+We wrote about Rust 2018 [first in
+March](https://blog.rust-lang.org/2018/03/12/roadmap.html), [and then in
+July](https://blog.rust-lang.org/2018/07/27/what-is-rust-2018.html).
+For some more background about the *why* of Rust 2018, please go read those
+posts; there's a lot to cover in the release announcement, and so we're going
+to focus on the *what* here.
+
+Let's create a new project with Cargo:
+
+```console
+$ cargo new foo
+```
+
+Here's the contents of `Cargo.toml`:
+
+```toml
+[package]
+name = "foo"
+version = "0.1.0"
+authors = ["Your Name <you@example.com>"]
+edition = "2018"
+
+[dependencies]
+```
+
+A new key has been added under `[package]`: `edition`. You'll note it's been
+set to `2018`. You can also set it to `2015`, which is the default if the key
+does not exist.
+
+By using Rust 2018, you unlock some new features that are not allowed in Rust 2015.
+And eventually, we'll be turning on some new warnings that help you use these
+new features more effectively, as well.
+
+But it's important to note that each package you use can be in either 2015 or
+2018 mode, and they work seamlessly together. Your 2018 project can use 2015
+dependencies, and a 2015 project can use 2018 dependencies. This ensures that
+we don't split the ecosystem, and all of these new things are opt-in,
+preserving compatibility for existing code. Furthermore, when you do choose
+to migrate Rust 2015 code to Rust 2018, the changes can be made
+automatically, via `cargo fix`.
+
+What kind of new features, you may ask? Well, first, features get added to
+Rust 2015 unless they require some sort of incompatibility with 2015's
+features. As such, most of the language is available everywhere. You can
+check out [the edition
+guide](https://doc.rust-lang.org/nightly/edition-guide) to check each
+feature's minimum `rustc` version as well as edition requirements. However,
+there are a few big-ticket features we'd like to mention here: non-lexical
+lifetimes, and some module system improvements.
+
+#### Non-lexical lifetimes
+
+If you've been following Rust's development over the past few years, you may
+have heard the term "NLL" or "non-lexical lifetimes" thrown around. This is
+jargon, but it has a straightforward translation into simpler terms: the
+borrow checker has gotten smarter, and now accepts some valid code that it
+previously rejected. Consider this example:
+
+```rust
+fn main() {
```

Comment:
I wonder if we should use, or at least mention, a more real example? Maybe we should just leave it for @nikomatsakis's post, though.

Reply:
A "mutate the original variable at the end of a `match`" example might be good. IMO That's one of the biggest pain points. Perhaps this example from the RFC?

```
fn process_or_default() {
    let mut map = ...;
    let key = ...;
    match map.get_mut(&key) { // -------------+ 'lifetime
        Some(value) => process(value),     // |
        None => {                          // |
            map.insert(key, V::default()); // |
            //  ^~~~~~ ERROR.              // |
        }                                  // |
    } // <------------------------------------+
}
```

Reply:
I'd like to stick to the simple example for now, unless we feel *really* strongly about it. There's a lot in this post already, and keeping it straightforward is more important than being hyper realistic, I think.

If people feel super strongly I'm willing to change it though.

Reply:
I did link to the edition guide, which has the full details of everything; maybe we need another one?